contributors/design-proposals/raw-block-pv.md

```diff
@@ -0,0 +1,721 @@
+	// - spec: The v1.Volume spec
+	// - pod: The enclosing pod
+	NewBlockVolumeMapper(spec *Spec, podRef *v1.Pod, opts VolumeOptions) (BlockVolumeMapper, error)
+	// NewBlockVolumeUnmapper creates a new volume.BlockVolumeUnmapper from recoverable state.
+	// - name: The volume name, as per the v1.Volume spec.
+	// - podUID: The UID of the enclosing pod
+	NewBlockVolumeUnmapper(name string, podUID types.UID) (BlockVolumeUnmapper, error)
+	// ConstructBlockVolumeSpec constructs a volume spec based on the given
+	// pod name, volume name and a pod device map path.
+	// The spec may have incomplete information due to limited information
+	// from input. This function is used by volume manager to reconstruct
+	// volume spec by reading the volume directories from disk.
+	ConstructBlockVolumeSpec(podName, volumeName, mountPath string) (*Spec, error)
+```
+## Changes for volume mount points
+
+Currently, a volume which has filesystem is mounted to the following two paths on a kubelet node when the volumes is in-use.
+The purpose of those mount points are that Kubernetes manages volume attach/detach status using these mount points and number
+of references to these mount points.
+
+```
+- Global mount path
+/var/lib/kubelet/plugins/kubernetes.io/{pluginName}/{volumePluginDependentPath}/
+
+- Volume mount path
+/var/lib/kubelet/pods/{podUID}/volumes/{escapeQualifiedPluginName}/{volumeName}/
+```
+
+Even if the volumeMode is "Block", similar scheme is needed. However, the volume which 
+doesn't have filesystem can't be mounted.
+Therefore, instead of volume mount, we use symbolic link which is associated to raw block device.
+Kubelet creates a new symbolic link under the new `global map path` and `pod device map path`.
+
+#### Global map path for "Block" volumeMode volume
+Kubelet creates a new symbolic link under the new global map path when volume is attached to a Pod.
+Number of symbolic links are equal to the number of Pods which use the same volume. Kubelet needs
+to manage both creation and deletion of symbolic links under the global map path. The name of the
+symbolic link is same as pod uuid. Also this path contains plugin unique information like
+"500a0982991b8dc5-lun-0" to reconstruct volume spec from the path.
+```
+/var/lib/kubelet/plugins/kubernetes.io/{pluginName}/volumeDevices/{volumePluginDependentPath}/{pod uuid1}
+/var/lib/kubelet/plugins/kubernetes.io/{pluginName}/volumeDevices/{volumePluginDependentPath}/{pod uuid2}
+...
+```
+
+{volumePluginDependentPath} example:
+```
+FC plugin: {wwn}-lun-{lun} or {wwid}
+ex. /var/lib/kubelet/plugins/kubernetes.io/fc/volumeDevices/500a0982991b8dc5-lun-0/f527ca5b-6d87-11e5-aa7e-080027ff6387
+iSCSI plugin: {portal ip}-{iqn}-lun-{lun}
+ex. /var/lib/kubelet/plugins/kubernetes.io/iscsi/volumeDevices/1.2.3.4:3260-iqn.2001-04.com.example:storage.kube.sys1.xyz-lun-1/f527ca5b-6d87-11e5-aa7e-080027ff6387
+ ```
+ 
+#### Pod device map path for "Block" volumeMode volume
```

Comment:
What is the pod device map used for? Volume reconstruction?

Reply:
@msau42 
There are two usages.
- makeBlockVolumes(): to get block device path from symbolic link
- Volume reconstruction: during volume reconstruction, ConstructBlockVolumeSpec() uses destination of symbolic link under pod device map path to find corresponding symbolic link from global map path.

Reply:
Can you clarify the uses here?  The section above also mentions reconstruction, which is not correct.